m2cgen/interpreters/elixir/code_generator.py

```diff
@@ -0,0 +1,62 @@
+from contextlib import contextmanager
+
+from m2cgen.interpreters.code_generator import CodeTemplate, FunctionalCodeGenerator
+
+
+class ElixirCodeGenerator(FunctionalCodeGenerator):
+    tpl_function_signature = CodeTemplate("{function_name} = fn ->")
+    tpl_if_statement = CodeTemplate("cond do {if_def} -> ")
+    tpl_else_statement = CodeTemplate("true ->")
+    tpl_num_value = CodeTemplate("{value}")
+    tpl_block_termination = CodeTemplate("end")
+    tpl_infix_expression = CodeTemplate("({left}) {op} ({right})")
+    tpl_module_definition = CodeTemplate("""defmodule {module_name} do
+    @compile {{:inline, read: 2}}
+    defp read(bin, pos) do
+        <<_::size(pos)-unit(64)-binary, value::float, _::binary>> = bin
+        value
```

Comment:
It is a function, which will be inlined at compile-time.

This function receives a binary and a position, and returns the float which was at that position in the binary.

Elixir doesn't have arrays, and lists are not cheap to use, so the best solution which doesn't allocate a lot of memory and has O(1) lookup is using a binary.

The code in question is doing a pattern match in the binary to retrieve the value.


Say you have a binary (imagine it as an array of floats)

```
[0, 1, 2, 3, 4]
```

and you want to get item 2.

You create a pattern which looks like this:

```ex
<<_::size(2)-unit(64)-binary, value::float, _::binary>>
```

where the first part is:

```ex
_::size(2)-unit(64)-binary
```

which means: match first 2 items, and ignore the actual values

the second part is:
```ex
value::float
```

match a float and bind it's value to "value" name/variable

and the third part is:

```ex
_::binary
```

match anything that's binary, ignore the actual values


So, in the end, to get item 2, you create a mask like this:

```
[1,2,3,4,5]
[_, _, value, _, _]
``` 

and return the value.

This snippet was suggested by Elixir's creator, BTW: https://github.com/elixir-nx/nx/issues/435

Reply:
Awesome explanation, thanks!

From the linked issues and topics I got that working with binaries would be much faster. But what if we make `score()` function take and return lists for better user experience (I guess it's easier to work with lists instead of binaries for end-users)? Only `score()` function, not any other internal ones into which code is being split according to the usage of `BinExpressionDepthTrackingMixin`. Something like take a list and immediately convert it to a binary, and similarly convert a final binary to a list just before returning from `score()` function. WDYT about it? I hope it won't hurt efficiency much but greatly ease the usage of generated code for users.

Reply:
I believe we should export 2 versions, then.

The list is bad in situations like batch predictions, where you could avoid allocating the list, and there is no way to avoid paying for that...

By returning the binary, we have a performant version with no performance drawbacks, but which is harder to use...

Reply:
Unfortunately m2cgen was not designed to handle batch predictions. So I think by default we can work with lists to ease user experience. As an option, we can add new bool argument (default `False`) which will control whether to work with binaries end-to-end. Or what did you mean by
> we should export 2 versions
?

Reply:
Yeah, that was what I meant by "export 2 versions". Receiving the new bool argument could work too.

I'll implement it, sorry for the delay - rough days on my job :)

Reply:
In the end I preferred not having the 2 versions, or the bool setting.

I think that returning and receiving lists is a good approach.

It was implemented in https://github.com/lucasavila00/m2cgen/commit/f27e4cafbd62c78bee12846cf3e070ea907befa9

Reply:
Got it, thanks a lot!